src/components/[guild]/Requirements/components/GuildCheckout/hooks/usePurchaseAsset.ts

```diff
@@ -0,0 +1,192 @@
+) => {
+  const { chainId, tokenBuyerContract } = data
+
+  if (!purchaseSupportedChains.ERC20?.includes(Chains[chainId]))
+    return Promise.reject("Unsupported chain")
+
+  const getAssetsParams = generateGetAssetsParams(data)
+
+  try {
+    await tokenBuyerContract.callStatic.getAssets(...getAssetsParams)
+  } catch (callStaticError) {
+    if (callStaticError.error) {
+      const walletError = processWalletError(callStaticError.error)
+      return Promise.reject(walletError.title)
+    }
+
+    if (!callStaticError.errorName) return Promise.reject(callStaticError)
+
+    // TODO: we could handle Uniswap universal router errors too
+    switch (callStaticError.errorName) {
+      case "AccessDenied":
+        return Promise.reject("TokenBuyer contract error: access denied")
+      case "TransferFailed":
+        return Promise.reject("TokenBuyer contract error: ERC20 transfer failed")
+    }
+  }
+
+  const getAssetsCall = await tokenBuyerContract.getAssets(...getAssetsParams)
+
+  setTxHash(getAssetsCall.hash)
+
+  return getAssetsCall.wait()
+}
+
+const usePurchaseAsset = () => {
+  const { addDatadogAction, addDatadogError } = useDatadog()
+
+  const showErrorToast = useShowErrorToast()
+  const toast = useToast()
+
+  const { account, chainId } = useWeb3React()
+
+  const {
+    requirement,
+    pickedCurrency,
+    txHash,
+    setTxHash,
+    setTxSuccess,
+    setTxError,
+  } = useGuildCheckoutContext()
+  const {
+    data: { symbol },
+  } = useTokenData(requirement.chain, requirement.address)
+  const {
+    data: {
+      priceInWei,
+      guildFeeInWei: rawGuildFeeInWei,
+      buyAmountInWei,
+      source,
+      path,
+      tokenAddressPath,
+    },
+  } = usePrice(pickedCurrency)
+
+  const tokenBuyerContract = useContract(TOKEN_BUYER_CONTRACT, TOKEN_BUYER_ABI, true)
+
+  const purchaseAssetData = useMemo(() => {
+    if (
+      !chainId ||
+      !account ||
+      !tokenBuyerContract ||
+      !pickedCurrency ||
+      !priceInWei ||
+      !rawGuildFeeInWei ||
+      !buyAmountInWei ||
+      !source ||
+      (!path && !tokenAddressPath)
+    )
+      return undefined
+
+    const amountIn = BigNumber.from(priceInWei)
+    const guildFeeInWei = BigNumber.from(rawGuildFeeInWei)
+    const amountInWithFee = amountIn.add(guildFeeInWei)
+    const amountOut = BigNumber.from(buyAmountInWei)
+
+    return {
+      chainId,
+      account,
+      tokenBuyerContract,
+      tokenAddress: pickedCurrency,
+      amountIn,
+      amountInWithFee,
+      amountOut,
+      source,
+      path,
+      tokenAddressPath,
+    }
+  }, [
+    chainId,
+    account,
+    tokenBuyerContract,
+    pickedCurrency,
+    priceInWei,
+    rawGuildFeeInWei,
+    buyAmountInWei,
+    source,
+    path,
+    tokenAddressPath,
+  ])
+
+  const generatedGetAssetsParams = useMemo(
+    () =>
+      purchaseAssetData ? generateGetAssetsParams(purchaseAssetData) : undefined,
+    [purchaseAssetData]
+  )
```

Comment:
Can't we move calculating these to happen later, in the core functions? So we don't need these useMemos. As we already call `generateGetAssetsParams` in purchaseAsset, but pre-calculate it for useEstimateGasFee here. Ideally we'd just pass a purchaseAssetData containing variables in their original form, and convert them to BigNumber / whatever we need in the core functions with an abstracted function like generateGetAssetsParams

Reply:
@dovalid I ended up moving this logic (L92-L134) into the `generateGetAssetsParams` util function (see here f58e24a8b6fe6731d0336fe50c713a9824f9faca) and calling only that function inside the `usePurchaseAsset` hook, since we actually need to pass the generated params to the `useEstimateGasFee` hook (it's a general hook, which can be used with any contract, so I didn't want to put any TokenBuyer-specific logic there).

Let me know if you still think that we should avoid calling the `generateGetAssetsParams` function here.

Reply:
It's perfect this way thanks!